File: app/src/main/java/io/opentakserver/opentakicu/presets/ServerPreset.java
package io.opentakserver.opentakicu.presets;

import androidx.annotation.NonNull;

import org.json.JSONException;
import org.json.JSONObject;

/**
 * One saved streaming destination. Holds just the connection fields (not codec/bitrate) so the
 * operator can keep several servers — home MediaMTX, a field relay, a teammate's laptop — and
 * switch between them in a single tap from Streaming settings.
 */
public class ServerPreset {
    public String name;
    public String protocol;
    public String address;
    public String port;
    public String path;
    public String username;
    public String password;
    public boolean tcp;
    /** Viewer/playback URL sent to ATAK via CoT (may differ from the publish URL). Blank = auto. */
    public String observerUrl;

    public ServerPreset() {}

    public ServerPreset(String name, String protocol, String address, String port,
                        String path, String username, String password, boolean tcp, String observerUrl) {
        this.name = name;
        this.protocol = protocol;
        this.address = address;
        this.port = port;
        this.path = path;
        this.username = username;
        this.password = password;
        this.tcp = tcp;
        this.observerUrl = observerUrl;
    }

    /**
     * Suggest a sensible viewer/observer URL from the connection fields. RTSP/RTMP viewer == publish
     * URL (RTSP gets a trailing {@code ?tcp}); SRT viewers use {@code streamid=read:NAME}.
     */
    public static String suggestObserverUrl(String protocol, String address, String port, String path) {
        String a = address == null ? "" : address.trim();
        String p = path == null ? "" : path.trim();
        String pt = port == null ? "" : port.trim();
        String proto = protocol == null ? "rtsp" : protocol.trim().toLowerCase();
        switch (proto) {
            case "rtmp":
            case "rtmps":
                return proto + "://" + a + ":" + (pt.isEmpty() ? "1935" : pt) + "/" + p;
            case "srt":
                return "srt://" + a + ":" + pt + "?streamid=read:" + p;
            case "udp":
                return "udp://" + a + ":" + pt;
            case "rtsp":
            case "rtsps":
            default:
                return proto + "://" + a + ":" + (pt.isEmpty() ? "8554" : pt) + "/" + p + "?tcp";
        }
    }

    /** A one-line "rtsp://host:port/path" summary for the list UI. */
    @NonNull
    public String summary() {
        String p = protocol == null ? "" : protocol;
        String a = address == null ? "" : address;
        String pt = port == null ? "" : port;
        String pa = path == null ? "" : path;
        return p + "://" + a + ":" + pt + "/" + pa;
    }

    public JSONObject toJson() throws JSONException {
        JSONObject o = new JSONObject();
        o.put("name", name);
        o.put("protocol", protocol);
        o.put("address", address);
        o.put("port", port);
        o.put("path", path);
        o.put("username", username);
        o.put("password", password);
        o.put("tcp", tcp);
        o.put("observerUrl", observerUrl);
        return o;
    }

    public static ServerPreset fromJson(JSONObject o) {
        ServerPreset p = new ServerPreset();
        p.name = o.optString("name", "");
        p.protocol = o.optString("protocol", "rtsp");
        p.address = o.optString("address", "");
        p.port = o.optString("port", "8554");
        p.path = o.optString("path", "");
        p.username = o.optString("username", "");
        p.password = o.optString("password", "");
        p.tcp = o.optBoolean("tcp", false);
        p.observerUrl = o.optString("observerUrl", "");
        return p;
    }
}

File: ios/Sentinel/Models/ServerPreset.swift
//
//  ServerPreset.swift
//  Mirrors the Android model exactly so JSON exports are interchangeable across platforms.
//

import Foundation

/// A saved streaming destination. Two URL surfaces:
///  • **Publish** — protocol/address/port/path/user/pass/tcp that SENTINEL pushes to.
///  • **Viewer**  — the URL advertised to ATAK in the CoT `<__video>` (peers tap → watch). May
///    differ from publish (SRT `read:` vs `publish:`, RTSP `?tcp` for forced TCP playback, etc.).
struct ServerPreset: Identifiable, Hashable, Codable {
    var id: UUID = .init()
    var name: String = ""
    var protocolName: String = "rtsp"       // rtsp, rtsps, rtmp, rtmps, srt, udp
    var address: String = ""
    var port: String = "8554"
    var path: String = ""
    var username: String = ""
    var password: String = ""
    var tcp: Bool = false
    /// Viewer URL sent to ATAK in CoT. Blank → auto-derive from the publish fields.
    var observerURL: String = ""

    enum CodingKeys: String, CodingKey {
        case id, name
        case protocolName = "protocol"
        case address, port, path, username, password, tcp, observerURL
    }

    /// One-line summary used in the preset list.
    var summary: String { "\(protocolName)://\(address):\(port)/\(path)" }

    /// Auto-derive a sensible viewer URL from the publish fields. Mirrors the Android
    /// `ServerPreset.suggestObserverUrl` behaviour exactly.
    static func suggestObserverURL(protocolName: String, address: String, port: String, path: String) -> String {
        let proto = protocolName.lowercased()
        switch proto {
        case "rtmp", "rtmps":
            return "\(proto)://\(address):\(port.isEmpty ? "1935" : port)/\(path)"
        case "srt":
            return "srt://\(address):\(port)?streamid=read:\(path)"
        case "udp":
            return "udp://\(address):\(port)"
        default:    // rtsp / rtsps
            return "\(proto)://\(address):\(port.isEmpty ? "8554" : port)/\(path)?tcp"
        }
    }
}

/// MediaMTX default ingest ports per protocol — used to auto-fill the Port field on protocol
/// change in the editor.
enum DefaultPorts {
    static func forProtocol(_ p: String) -> String {
        switch p.lowercased() {
        case "rtmp", "rtmps": return "1935"
        case "srt":           return "8890"
        case "udp":           return "1234"
        default:              return "8554"  // rtsp / rtsps
        }
    }
    static let knownDefaults: Set<String> = ["8554", "1935", "8890", "1234"]
}

/// Protocol pickers in the UI. `publishProtocols` matches what SENTINEL can push;
/// `viewerProtocols` is what ATAK can play (URLs only — no rtsps/rtmps separate from rtsp/rtmp).
enum StreamProtocols {
    static let publish: [String] = ["rtsp", "rtsps", "rtmp", "rtmps", "srt", "udp"]
    static let viewer:  [String] = ["rtsp", "rtmp", "srt"]

    /// Map a publish protocol to the playable viewer protocol ATAK should default to.
    static func mapToViewer(_ publish: String) -> String {
        switch publish.lowercased() {
        case "srt":            return "srt"
        case "rtmp", "rtmps":  return "rtmp"
        default:               return "rtsp"
        }
    }
}
